### ISSUE TITLE
Error handling

### ISSUE THREAD

**Issue by joakim-hove** [OPM/opm-common#1934]
A project has been started to improve error handling in flow. This is a large task and help and help in pointing out bad messages and bad handling of errors would be appreciated. I suggest people add a comment here with their favorite poor error message. As I see it there are at least two axes for improvement in the error handling:

**Better messages:** This is quite "simple" - improve messages all over the place, in particular make them more user centric and less developer centric. I suppose we start with messages which actually have hit the users - and then take from there. 

**Flow of error handling:** As illustrated [here](https://github.com/OPM/opm-common/pull/1931) an error can jump through many frames before the program is eventually terminated. What is the best way to handle this is not clear to me.

**Comment by joakim-hove** [comment]
**ParseContext: default exit?**
Much of the jumping about illustrated in https://github.com/OPM/opm-common/pull/1931 comes because the ParseContext default error handler is to throw an exception. If we instead set the default behavior to be `EXIT1` we would print one nicely formatted error message - and go home cleanly with `std::exit(1)`. That is less comfortable for developers - but cleaner output for users. I suggest this path.

**Comment by atgeirr** [comment]
> we would print one nicely formatted error message - and go home cleanly with `std::exit(1)`

I do not think exit() is very clean... But I agree we should not print 3 error messages for the same error condition. I think using exceptions to exit scopes at error is good, but the level where an error message is printed should be chosen to be the one where it is most useful for the user. In the example you linked, the first one is good and sufficient.

**Comment by joakim-hove** [comment]
> I do not think exit() is very clean...

At first glance I agree with you - and some time ago I would have been willing to fight hard for the exception, but now I have put on my user underwear and changed my mind:

1. We generally do *not* handle exceptions - they just bubble up - gradually loosing quality of context until the whole application stops with a less than informative message.

2. This will be configured in the `ParseContext.cpp` class - and only error conditions which are well understood and have a good error message will be configured to `std::exit(1)`. General exceptions will still bubble as they do now.

> But I agree we should not print 3 error messages for the same error condition.

The problem here is that it varies a lot how much context/information we have at the error site. In the situation illustrated in #1931 we have good information at error site, there is no need to curry the error messages further when backtracing the exception - on the other hand if the original problem is just: `throw std::out_of _range("Index too large");` - it is certainly valuable to add extra context information like which file and line we were currently working at.

**Comment by atgeirr** [comment]
I guess some of it comes down to this: when at a catch site, we presumably catch either because this is a good place and context to make a good error message, or because we can do something to handle the problem and continue. When we are dealing with conditions that should cause termination, there should probably never be more than a single catch statement, and that should do the logging. I know we break that rule...

**Comment by joakim-hove** [comment]
> just using std::exit will deadlock in parallel.

OK - I was afraid of that; it is not possible with some sort of heartbeat pattern - where the remaining N - 1 processes check the status of rank 0?

**Comment by joakim-hove** [comment]
**Custom exception class: OpmError ?**

I think the "jumping about" illustrated in #1931 is problematic. One possible suggestion would be that we implement a custom exception - `OpmError` - the *convention* around `OpmError` could  be the following:

1. `OpmError` is only created in situations where we have all the necessary context and can create a good error message. 
2. When an `OpmError` is created the message is also logged - no need for separate logging of `exc.what()` later.
3. An `OpmError` is not caught and rethrown.

This will ensure that a well understood error condition like the one illustrated in #1931 only gets one display on the terminal/log, a more generic `std::out_of_range("Invalid index ...")` will still be treated roughly the way it is now.

**Comment by blattms** [comment]
You can issue  `MPI_Abort(MPI_COMM_WORLD)` and hope that the MPI implementation is decent, but it is not nice.

I am a bit surprised that we are now opting for a hard exit when an error happens. Wasn't the current behaviour to parse everything and present the errors at the end a deliberate decision for usability. The user gets present all errors and warnings instead just the first one, and can fix them. Instead of running flow, get an error, fix it, run flow, get an error, fix it, ...

**Comment by joakim-hove** [comment]
> I am a bit surprised that we are now opting for a hard exit when an error happens. Wasn't the current behaviour to parse everything and present the errors at the end a deliberate decision for usability

Yes you are right - I back down; I am just a frustrated that the downstream error handling is difficult and we have not (yet at least) been able to benefit from this.

**Comment by joakim-hove** [comment]
![image](https://user-images.githubusercontent.com/1321665/93870614-c31a5700-fccd-11ea-9e77-996b245acc69.png)

**Comment by alfbr** [comment]
While it is still a bit early to put a straight jacket on error messages (since there may be many corner cases we are unaware of at this point), a pattern starts to stand out. That is the following:

Line one: Message outlining the error with reference to keyword, and a table number if relevant.
Line two: In /path/to/file line 23
Line three: A helpful message to the user informing of problems that may arise or possible work-arounds

Line two mighet expand to multiple lines if it is a recurring item in a table, please look at https://github.com/OPM/opm-common/pull/1938 for suggested handling.

Line three may also be a multi-line message if warranted.

All opinions on the matter is highly appreciated.

**Comment by alfbr** [comment]
During a discussion in https://github.com/OPM/opm-common/pull/1939 the question about indenting similarly to Eclipse came up. I have to admit that I am still uncertain about whether we want indenting, but somewhat related to that is the fact that colour coding of messages is lost when terminal output is piped or redirected to file. However, the escape characters from the colouring is present causing visual noise. If I understand the following link correct (see the answer):
https://askubuntu.com/questions/1214517/c-colored-output-text-when-redirected-in-text-file
it is possible to have Flow remove the colour coding whenever the terminal output is piped or redirected to a file. If so, I believe it should be done. Moreover, it allows for the possibility to have indenting only happen when output is piped or redirected to a file also. In this case, after discussion with my two best colleagues, I suggest the indenting is only two characters, as that is sufficient to make it stand out. Using ten characters would cause excessive line breaks in too many cases.

**Comment by OPMUSER** [comment]
@atgeirr send me a REGEX expression to filter out these escape code, and the intent was to parse the LOG file in OPMRUN to filter these out. The problem is my to do list keeps on growing and I need get OPMRUN formulated for pip first.

I have a strong preference to indent like this in the print file:
```

Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
         In file /home/hove/work/OPM/opm-tests/spe1/SPE1CASE1.DATA at line 253.  
         Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  

Warning: Keyword OPTIONS not implemented in OPM Flow

```
For greater readability.

**Comment by OPMUSER** [comment]
I'm user not a developer, so this should carry much more wait :-)

**Comment by alfbr** [comment]
> @atgeirr send me a REGEX expression to filter out these escape code, and the intent was to parse the LOG file in OPMRUN to filter these out. The problem is my to do list keeps on growing and I need get OPMRUN formulated for pip first.

Unfortunately, we cannot expect all users to get cross that bridge. Do you agree that having Flow strip it automatically when you redirect output is something we should do? If so, you need do nothing in opmrun, it will all be done automatically.

> I have a strong preference to indent like this in the print file:
> 
> ```
> 
> Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
>          In file /home/hove/work/OPM/opm-tests/spe1/SPE1CASE1.DATA at line 253.  
>          Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  
> 
> Warning: Keyword OPTIONS not implemented in OPM Flow
> ```
> 
> For greater readability.

I understand, and thanks for being clear about your preference. In my case the message would typically not look like that though. It would rather look like this (or worse):

```
 Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
          In file /home/hove/work/OPM/opm-tests/spe1-case/my/path/is/never/this/short/SPE1CASE1.DATA
          at line 253.
          Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  

```
A warning would eat up even more characters from the available line. Please have another look at the messages from Eclipse. As a consequence they are typically very short and often not that helpful. If we go with a couple of characters indent when redirecting to file, it would look like this:
```
Warning: VFP table 6 has BHP versus THP not increasing in keyword VFPPROD.
  In /home/myusername/Iteration-1/realization-3/eclipse/include/INCLUDE/VFP/GAS_PD2.VFP line 62
  This may cause convergence issues due to switching between BHP and THP control.

Warning: VFP table 3 has BHP versus THP not increasing in keyword VFPPROD.
  In /home//myusername/Iteration-1/realization-3/eclipse/include/INCLUDE/VFP/AlmostVertNew.VFP line 9
  This may cause convergence issues due to switching between BHP and THP control.
```

**Comment by alfbr** [comment]
Actually, unless we let the output machinery break lines, it would look like this:
```
 Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
          In file /home/hove/work/OPM/opm-tests/spe1-case/my/path/is/never/this/short/SPE1CASE1.DATA
at line 253.
          Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  

```

**Comment by OPMUSER** [comment]
> Do you agree that having Flow strip it automatically when you redirect output is something we should do? If so, you need do nothing in opmrun, it will all be done automatically.

Yes, that is a more complete solution so +1 from me.

I'm okay with this as well, with the output machinery breaking the lines

```
 Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
          In file /home/hove/work/OPM/opm-tests/spe1-case/my/path/is/never/this/short/SPE1CASE1.DATA
          at line 253.
          Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  
```
But not this:
```

 Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
          In file /home/hove/work/OPM/opm-tests/spe1-case/my/path/is/never/this/short/SPE1CASE1.DATA
at line 253.
          Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  

```

**Comment by alfbr** [comment]
> I'm okay with this as well, with the output machinery breaking the lines

This is where I suspect you have not thought through the implications. Are you aware that you then suggest a limit on number of characters in a line, and its consequences for the totality? If so, you need at least to share at how many characters the line should break, and what rules it should break by (at first space after limit maybe?). It will otherwise be impossible to know what you actually are suggesting. You also need to share if you believe the right margin should be the same, I assume you will, but @atgeirr argued for shifting the right margin ten characters to the right for some messages in an earlier issue. By all means, you may still want this, I just want to make sure you see the implications of your suggestions.

**Comment by OPMUSER** [comment]
Some many questions, so little time :-)

- If we use the formatting I suggested, then I don't have a strong opinion on the line length 80/100/120/132. Although, the longer the better to capture the filename, as these seem to be getting longer and longer with each passing year. Also I assume that the number of lines per message is flexible and not limited to four in the examples. We should set set a limit of say 10 or something to cover more lengthy messages.

- Line break on first space after the line length limit.

- Right margin should be the same for each line,  except the first so that the messages look like this:
```
Error  : Record contains too many items in keyword EQUIL. Expected 9 items, found 18.
         In file /home/hove/work/OPM/opm-tests/spe1/SPE1CASE1.DATA at line 253.  
         Record is: "8400 4800 8450 0 8300 0 1 0 0 1 2 3 4 5 6 7 8 9"  

Warning: Keyword OPTIONS not implemented in OPM Flow
```

- Most editors can display 132 characters and wrapping can switched on or off, so I don't see line wrapping as an issue. Remember the reports on the print file go to ~132 and that is not an issue.

In summary, I have a strong preference for the format above, less concern on line length greater than or equal to 80, and that the number of lines in a message is flexible and not restricted to four as in the examples.
